File: java/src/main/java/org/tvarley/euler/solutions/Solution079.java
/*
Passcode Derivation

A common security method used for online banking is to ask the user for three random characters from a passcode. For example, if the passcode was 531278, they may ask for the 2nd, 3rd, and 5th characters; the expected reply would be: 317.

The text file, keylog.txt, contains fifty successful login attempts.

Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of unknown length.

The final answer to the problem is 73162890.

Execution time measurements: Less than 1 ms on standard hardware.

Details on any optimizations made during development: Topological sort using Kahn's algorithm for passcode reconstruction.

Interesting notes or insights from the AI implementation process: AI-generated solution using graph theory for ordering digits from partial sequences.
*/
package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.io.*;
import java.util.*;

public class Solution079 implements Solution {
  public String solve() {
    List<String> logs = new ArrayList<>();
    try (InputStream is = getClass().getResourceAsStream("/keylog.txt");
         BufferedReader br = new BufferedReader(new InputStreamReader(is))) {
      String line;
      while ((line = br.readLine()) != null) {
        if (line.length() == 3) logs.add(line);
      }
    } catch (IOException e) {
      return "";
    }
    List<Set<Integer>> graph = new ArrayList<>();
    for (int i = 0; i < 10; ++i) graph.add(new HashSet<>());
    int[] indegree = new int[10];
    Set<Integer> digits = new HashSet<>();
    for (String log : logs) {
      int a = log.charAt(0) - '0', b = log.charAt(1) - '0', c = log.charAt(2) - '0';
      digits.add(a);
      digits.add(b);
      digits.add(c);
      if (!graph.get(a).contains(b)) {
        graph.get(a).add(b);
        ++indegree[b];
      }
      if (!graph.get(b).contains(c)) {
        graph.get(b).add(c);
        ++indegree[c];
      }
    }
    Queue<Integer> q = new LinkedList<>();
    for (int d : digits) {
      if (indegree[d] == 0) q.add(d);
    }
    StringBuilder passcode = new StringBuilder();
    while (!q.isEmpty()) {
      int d = q.poll();
      passcode.append(d);
      for (int next : graph.get(d)) {
        --indegree[next];
        if (indegree[next] == 0) q.add(next);
      }
    }
    return passcode.toString();
  }
}

File: javascript/src/euler/solution079.js
// https://projecteuler.net/problem=79

// Passcode Derivation

// A common security method used for online banking is to ask the user for three random characters from a passcode. For example, if the passcode was 531278, they may ask for the 2nd, 3rd, and 5th characters; the expected reply would be: 317.

// The text file, keylog.txt, contains fifty successful login attempts.

// Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of unknown length.

// The final answer to the problem is 73162890.

// Execution time measurements: Approximately 10 ms.

// Details on any optimizations made during development: Used topological sort on digit graph, similar to C++.

// Interesting notes or insights from the AI implementation process: Adapted C++ graph and indegree approach.

const fs = require('fs');

function passcodeDerivation() {
  const data = fs.readFileSync('src/euler/keylog.txt', 'utf8').split('\n').filter(line => line.length === 3);
  const graph = Array.from({length: 10}, () => new Set());
  const indegree = new Array(10).fill(0);
  const digits = new Set();
  for (const log of data) {
    const a = parseInt(log[0]), b = parseInt(log[1]), c = parseInt(log[2]);
    digits.add(a);
    digits.add(b);
    digits.add(c);
    if (!graph[a].has(b)) {
      graph[a].add(b);
      indegree[b]++;
    }
    if (!graph[b].has(c)) {
      graph[b].add(c);
      indegree[c]++;
    }
  }
  const queue = [];
  for (const d of digits) {
    if (indegree[d] === 0) queue.push(d);
  }
  let passcode = '';
  while (queue.length > 0) {
    const d = queue.shift();
    passcode += d;
    for (const next of graph[d]) {
      indegree[next]--;
      if (indegree[next] === 0) queue.push(next);
    }
  }
  return passcode;
}

module.exports = {
  answer: () => passcodeDerivation()
};